Migrate the quantizer to use aten ops directly

mcremon-meta opened the pull request:
Summary:
This major change allows a lot more flexibility in the quantizer, and reduces the dependency on the decompositions/graph tracing tools.

The motivation is that some of those do not preserve or propagate `source_fn_stack` information, resulting in quantization misses. SDPA is an example, where the underlying `bmm` ops cannot be quantized with `source_fn_stack` information alone, or MHA, which can hide its SDPA component and sometimes even `linear` ops depending on the model (see ViT for an example).

Summary of the changes:
- change the quantizer to match aten ops directly, through `node.target`
- propagate required changes to the `QuantFusion` pass
- update/remove existing patterns

Differential Revision: D59552606

dulinriley commented on backends/cadence/aot/compiler.py:
```diff
@@ -53,8 +57,10 @@ def quantize_pt2(
-    # pyre-fixme[16]: Pyre doesn't get that CadenceQuantizer has a patterns attribute
-    patterns = [q.pattern for q in quantizer.quantizers]
+    patterns = [
+        assert_is_instance(q, CadenceAtenQuantizer).pattern
```
Is this a duplicate and not properly stacked with 
https://github.com/pytorch/executorch/pull/4047?

You can use "gh-stack" to help with this in the future

dulinriley commented on backends/cadence/aot/quantizer/patterns.py:
```diff
@@ -44,18 +47,20 @@ class PartitionAnchors:
 class QuantizationPattern(ABC):
     @abstractmethod
-    def partition_types(self):
+    def partition_types(self) -> list[OpOverload]:
```
I think we need to support Python 3.8 here, which doesn't support `list[x]`, and you need to use `typing.List[x]`

mcremon-meta committed:
Migrate the quantizer to use aten ops directly (#4195)

Summary:
Pull Request resolved: https://github.com/pytorch/executorch/pull/4195

This major change allows a lot more flexibility in the quantizer, and reduces the dependency on the decompositions/graph tracing tools.

The motivation is that some of those do not preserve or propagate `source_fn_stack` information, resulting in quantization misses. SDPA is an example, where the underlying `bmm` ops cannot be quantized with `source_fn_stack` information alone, or MHA, which can hide its SDPA component and sometimes even `linear` ops depending on the model (see ViT for an example).

Also note than in most cases, we match single nodes anyway, with a 1-1 mapping between the op (either nn.Module or nn.functional) and the aten op, so using the aten op directly is simply easier.

Summary of the changes:
- change the quantizer to match aten ops directly, through `node.target`
- propagate required changes to the `QuantFusion` pass
- update/remove existing patterns

Reviewed By: dulinriley

Differential Revision: D59552606
